assembly/src/sema/mod.rs

```diff
@@ -0,0 +1,209 @@
+                        span: import.span(),
+                    });
+                } else {
+                    analyzer.define_import(import)?;
+                }
+            }
+            Form::Procedure(export @ Export::Alias(_)) => match kind {
+                ModuleKind::Kernel => {
+                    docs.take();
+                    analyzer.error(SemanticAnalysisError::ImportToKernel {
+                        span: export.span(),
+                    });
+                }
+                ModuleKind::Executable => {
+                    docs.take();
+                    analyzer.error(SemanticAnalysisError::UnexpectedExport {
+                        span: export.span(),
+                    });
+                }
+                ModuleKind::Library => {
+                    analyzer.define_procedure(export.with_docs(docs.take()))?;
+                }
+            },
+            Form::Procedure(export) => match kind {
+                ModuleKind::Executable
+                    if export.visibility().is_exported() && !export.is_main() =>
+                {
+                    docs.take();
+                    analyzer.error(SemanticAnalysisError::UnexpectedExport {
+                        span: export.span(),
+                    });
+                }
+                _ => {
+                    analyzer.define_procedure(export.with_docs(docs.take()))?;
+                }
+            },
+            Form::Begin(body) if matches!(kind, ModuleKind::Executable) => {
+                let docs = docs.take();
+                let source_file = analyzer.source_file();
+                let procedure =
+                    Procedure::new(body.span(), Visibility::Public, ProcedureName::main(), 0, body)
+                        .with_docs(docs)
+                        .with_source_file(Some(source_file));
+                analyzer.define_procedure(Export::Procedure(procedure))?;
+            }
+            Form::Begin(body) => {
+                docs.take();
+                analyzer.error(SemanticAnalysisError::UnexpectedEntrypoint { span: body.span() });
+            }
+        }
+    }
+
+    if let Some(unused) = docs.take() {
+        analyzer.error(SemanticAnalysisError::UnusedDocstring {
+            span: unused.span(),
+        });
+    }
+
+    if matches!(kind, ModuleKind::Executable) && !analyzer.module().has_entrypoint() {
+        analyzer.error(SemanticAnalysisError::MissingEntrypoint);
+    }
+
+    if analyzer.has_errors() {
+        return analyzer.into_result();
+    }
+
+    // Run procedure checks
+    let module = visit_procedures(&mut analyzer)?;
+
+    // Check unused imports
+    for import in module.imports() {
+        if !import.is_used() {
+            analyzer.error(SemanticAnalysisError::UnusedImport {
+                span: import.span(),
+            });
+        }
+    }
+
+    if analyzer.has_errors() {
+        analyzer.into_result()
+    } else {
+        Ok(module)
+    }
```

Comment:
Why don't we use here just `analyzer.into_result()`? Is it because `into_result()` takes the module, and we want just to return a pointer (`Box`) to it?
Why can't we change `into_result()` to return a pointer? Is it correct that we want to have `into_result()` function with `take` because of some later implementations, which will use this behavior?

Reply:
Ah, you just made me realize I forgot to finish implementing support for warnings here properly! In any case, there is a reason for this (that now I have to actually realize, but expect that to happen tomorrow):

It's a slight artifact of how this PR has to represent warning diagnostics. The `miette` crate which provides the core infrastructure for diagnostics, requires types that implement `Diagnostic` to also implement `std::error::Error` (or the libcore equivalent). Despite being implemented as errors, it is the `severity` of the diagnostic that determines whether it is _really_ an error.

In the semantic analyzer, we gather all errors/warnings into a single `SyntaxError`, so that those errors/warnings are all emitted as a batch. However, we don't want to cause compilation to fail just because of warnings, so we are specifically checking here to see if there are any "real" errors, and if so, producing an `Err`, otherwise we will return `Ok`. In the "only warnings" case though, we will still output the warnings to stderr, we just won't return a `SyntaxError`.

The part I didn't finish here is:

1.) Distinguish between errors and warnings in the `has_errors` check
2.) Print warnings to stderr when there are warnings but no errors

I'll get to that tomorrow.